src/rust/engine/remote_provider/remote_provider_reapi/src/byte_store.rs

```diff
@@ -390,6 +444,139 @@ impl ByteStoreProvider for Provider {
         .map_err(|e| e.to_string())
     }
 
+    fn batch_load_supported(&self) -> bool {
+        self.batch_load_enabled
+    }
+
+    async fn load_batch(
+        &self,
+        digests: Vec<Digest>,
+        destination: &mut dyn BatchLoadDestination,
+    ) -> Result<HashMap<Digest, Result<bool, String>>, String> {
+        let mut max_batch_total_size_bytes = {
+            let capabilities = self.get_capabilities().await.map_err(|e| e.to_string())?;
+
+            capabilities
+                .cache_capabilities
+                .as_ref()
+                .map(|c| c.max_batch_total_size_bytes as usize)
+                .unwrap_or_default()
+        };
+        if max_batch_total_size_bytes == 0 || self.batch_api_size_limit < max_batch_total_size_bytes
+        {
+            max_batch_total_size_bytes = self.batch_api_size_limit;
+        }
+        if max_batch_total_size_bytes == 0 {
+            max_batch_total_size_bytes = DEFAULT_MAX_GRPC_MESSAGE_SIZE;
+        }
+        max_batch_total_size_bytes = max_batch_total_size_bytes
+            - self
+                .instance_name
+                .as_ref()
+                .cloned()
+                .unwrap_or_default()
+                .len()
+            - 10;
+
+        let mut chunks: Vec<Vec<Digest>> = Vec::new();
+        let mut current_chunk: Vec<Digest> = Vec::new();
+        let mut current_size = 0;
+
+        for digest in digests.iter() {
+            let message_size = digest.size_bytes + RPC_RESPONSE_PER_ITEM_SIZE;
+            if current_size + message_size >= max_batch_total_size_bytes {
+                chunks.push(std::mem::take(&mut current_chunk));
+                current_size = 0;
+            }
+            current_chunk.push(*digest);
+            current_size += message_size;
+        }
+
+        if !current_chunk.is_empty() {
+            chunks.push(current_chunk);
+        }
+
+        let client = self.cas_client.as_ref();
+        let destination = Arc::new(Mutex::new(destination));
+
+        let chunk_futures = chunks
+            .clone()
+            .into_iter()
+            .map(|chunk| {
+                let digests = chunk
+                    .iter()
+                    .map(|d| d.into())
+                    .collect::<Vec<remexec::Digest>>();
+                let request = remexec::BatchReadBlobsRequest {
+                    instance_name: self.instance_name.as_ref().cloned().unwrap_or_default(),
+                    digests: digests,
+                    acceptable_compressors: vec![],
+                };
+
+                let client = client.clone();
+                let destination = destination.clone();
+
+                retry_call(
```

Comment:
the currently `retry_call` makes this implementation slightly awkward since I can only return a single status for the retry check (I just return the first) as we cannot communicate that some of the items succeeded, but others failed. In practice this is probably fine? It just means the entire batch will be retried as a unit

Reply:
The main alternative I can think of is to remove `retry_call` and separate the successful ones from the failures, then make a smaller request for the failures. In practice I'm not sure how often this will be used, so I think I'd prefer to leave better retry handling to a future pr.

Reply:
Sounds good